src/main/java/org/apache/sling/discovery/oak/SlingIdCleanupTask.java

```diff
@@ -0,0 +1,484 @@
+
+            final Resource clusterInstances = ResourceHelper.getOrCreateResource(resolver,
+                    localConfig.getClusterInstancesPath());
+            final Resource idMap = ResourceHelper.getOrCreateResource(resolver,
+                    localConfig.getIdMapPath());
+            final Resource syncTokens = ResourceHelper.getOrCreateResource(resolver,
+                    localConfig.getSyncTokenPath());
+            resolver.revert();
+            resolver.refresh();
+
+            final ValueMap idMapMap = idMap.adaptTo(ValueMap.class);
+            final ModifiableValueMap syncTokenMap = syncTokens
+                    .adaptTo(ModifiableValueMap.class);
+            final Calendar now = Calendar.getInstance();
+            int removed = 0;
+            boolean mightHaveMore = false;
+            int localBatchSize = batchSize;
+            long localMinCreationAgeMillis = minCreationAgeMillis;
+            for (Resource resource : clusterInstances.getChildren()) {
+                if (!hasTopology || currentView != localCurrentView
+                        || !localCurrentView.isCurrent()) {
+                    // we got interrupted during cleanup
+                    // let's not commit at all then
+                    logger.info(
+                            "cleanup : topology changing during cleanup - not committing this time - stopping for now.");
+                    return true;
+                }
+                final String slingId = resource.getName();
+                logger.info("cleanup : handling slingId = {}", slingId);
+                Object clusterNodeId = idMapMap.get(slingId);
+                if (clusterNodeId == null) {
+                    logger.info("cleanup : slingId not recently in use : {}",
+                            clusterNodeId);
+                } else {
+                    logger.info("cleanup : slingId WAS recently in use : {}",
+                            clusterNodeId);
+                    continue;
+                }
+                if (activeSlingIds.contains(slingId)) {
+                    logger.info("cleanup : slingId is currently active : {}", slingId);
+                    continue;
+                }
+                if (deleteIfOldSlingId(resource, syncTokenMap, now,
+                        localMinCreationAgeMillis)) {
+                    if (++removed >= localBatchSize) {
+                        // we need to stop
+                        mightHaveMore = true;
+                        break;
+                    }
+                }
+            }
+            if (!hasTopology) {
+                // we got interrupted during cleanup
+                // let's not commit at all then
+                logger.info(
+                        "cleanup : topology changing during cleanup - not committing this time - stopping for now.");
+                return true;
+            }
+            if (removed > 0) {
```

Comment:
```suggestion
            if (resolver.hasChanges > 0) {
```

Reply:
The goal with `removed` was to have a precise counter how many garbage slingIds were deleted, that's why it's tracked further up.